### DISCUSSION TITLE
NPM public packages can't be accessed without a token

### DISCUSSION THREAD

**Discussion by sergioisidoro** [community/community#33875] [Packages]
Container images can apparently be made public, and accessed anonymously: 

> A public package can be accessed anonymously without authentication. Once you make your package public, you cannot make your package private again.

https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility

However a NPM package cannon:
> You need an access token to publish, install, and delete private, internal, and public packages.

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry#authenticating-to-github-packages

This is quite confusing and counter intuitive. Why can't public NPM packages be accessed without authentication? Is this a problem of conflicting docs?  I've spent a considerable amount of time trying to install a package from the NPM Github registry on a public package, but constantly got an error about missing auth token.

Am I doing something wrong, or is this by design?

**Comment by mduenias** [comment]
I want to chime in as I am experiencing the same issue.

We are publishing our component library as an NPM package in the GitHub registry. 
This means all our consumer teams have to do the PAT set up locally and in their pipelines, which is even more complex when they don't use GitHub actions. So this configuration is hitting our adoption.

Having to configure the PAT hinder developer experience, specially when your package does not need to be private or requires authentication to be consumed.

I would be curious to know as well if this access gate is by design and is never going to change. So that we can take the decision to create an organisation in NPM and publish our package there.

Thanks so much!

**Comment by gerdstolpmann** [comment]
We just ran into the same issue - some of our our npm's are only containing glue code (which is quite common when you develop for WebAssembly and all the IP is in the wasm binary). It would be very useful to access these npm's without token.

**Comment by tofran** [comment]
> So that we can take the decision to create an organisation in NPM and publish our package there.

In reality it is way trickier:

1. If you are publishing a package in a private registry outside NPM, like the GitHub package registry you should have an organization with the same name (scope) in the official NPM registry. This is required to avoid dependency confusion attacks.
2. If you start publishing public packages in your NPM official registry (with the same scope as you private registry), you will encounter a problem: devices configured to point to the private registry will also try to fetch public packages form there (as you can only configure the registry per scope, not package). Meaning that you are forced to also publish the public packages to the registry being used by private packages (ex GitHub).
3. Now the worst part, if developers have this configuration on their computer and generate a `package-lock.json`, it will have the registry hardcoded there. If you make that repository public, anyone trying to `npm install` will not be able to, even if all the dependencies are public, because GitHub requires an authentication token. Basically you would have to regenerate the `package-lock.json`.

This is quite unfortunate, I think GitHub really needs to fix it.

Hopefully I will eventually write an in-depth blog post about this.

**Comment by TheKnarf** [comment]
Not being able to install public npm packages from the Github registry without creating a personal access token makes it entirely unusable for a lot of normal use-cases...

I wish they would fix this bug.

**Comment by mayank99** [comment]
With the recent release of [JSR](https://jsr.io/), I think this issue deserves some attention. JSR has some interesting ideas that GitHub Packages could consider borrowing if it wants to become a serious contender.

Obviously, JSR packages are publicly installable without auth. Still, JSR is an independent registry with different packages names from the NPM registry. For compatibility with NPM, all JSR packages are hosted under the `@jsr` scope. This is what enables the use of JSR registry only for JSR packages, without impacting other NPM packages. ([More details](https://jsr.io/docs/npm-compatibility#advanced-setup))

Another thing worth mentioning is the idea of ["upstream sources"](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/upstream-sources?view=azure-devops) which Azure Artifacts uses. If GitHub Packages wants to allow mixing-and-matching of packages from the same scope hosted on different registries, this could be a good solution.

**Comment by amber-beasley-liatrio** [comment]
This has posed as a problem for our org and we are now forced to publish to npmjs instead of the GitHub npm registry so packages can be truly public (accessible as anonymous).

**Comment by pablo-mayrgundter** [comment]
Agreed, this was not expected and I'm moving our projects off of github npm registry, back to npmjs

**Comment by tutods** [comment]
Any solution for this?

**Reply by jonespm** [reply]
Only solution I've used is to not use the package and just install a tag direct from Github with [github: syntax](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#repository)

`  "repository": "github:npm/example#v1.0.0",
`

**Reply by maddsua** [reply]
> Only solution I've used is to not use the package and just install a tag direct from Github with [github: syntax](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#repository)
> 
> ` "repository": "github:npm/example#v1.0.0",`

Wait, you could do this?

**Reply by jonespm** [reply]
Yeah, looks it's been a thing in npm for at _least_ 9 years, works fine. https://github.com/mozilla-magnet/magnet-client/blob/master/package.json

I think I'd originally found this on this other discussion here and hadn't used it before but works fine for some local packages. https://github.com/orgs/community/discussions/19037#discussioncomment-9579768

But would be nice if the regular package worked.

**Reply by ThePlenkov** [reply]
but that's not same - they you just reference the source code - and there may be some missing parts which are only present in publish CI..

**Comment by resistdesign** [comment]
This is amateur hour. +1

You should obviously not have to auth for public content ever.

That's why it's called...

**Comment by ThePlenkov** [comment]
It feels like I'm moving to [GItlab CI/CD for Github projects ](https://docs.gitlab.com/ci/ci_cd_for_external_repos/) Accroding to this page https://docs.gitlab.com/user/packages/npm_registry/#authenticate-to-the-package-registry 

> You must authenticate to the package registry to publish or install a package from a private project or a private group. You don’t need to authenticate if the project or the group is public.

That's the exact behavior I expect from Github NPM registry for public packages!

**Comment by cmahnke** [comment]
Maybe this discussion can be converted to a bug? Which repository would such a report to failed agains?

**Reply by maddsua** [reply]
This seems to be a conscious design decision on the GtiHub part, that affects every npm repo "published" there.

This means, you have to pass a PAT or an equivalent token around every time you wan to install a package from GH's registry in a CI environment that isn't Actions. The same holds true for local npm installs.

It is just another completely unnecessary spoke in the wheel. What makes it even more confusing, is that both GH and npmjs.org are now owned by MSFT. And somehow, npmjs.org doesn't require any auth in order to download packages from it.

At this point npmjs.org should be deprecated altogether in favor of using GH registry directly. Only after the current auth behavior is reconsidered, of course.

**Reply by cmahnke** [reply]
Yes, but even if it's a choice this wouldn't make it less a bug. I would expect that the package registry is public, which it currently isn't. As you said, the current way it's almost unusable when you don't control the complete environment. This works in corporate settings or an a CI system where you have a token configured. But it can e considered unusable if one is targeting reuse by third (unknown) parties.

Which one should be deprecated is hard to say, both have their short comings.
